editor/src/com/talosvfx/talos/editor/addons/scene/widgets/directoryview/DirectoryViewWidget.java

```diff
@@ -494,119 +496,221 @@ public void drop (DragAndDrop.Source source, DragAndDrop.Payload payload, float
-					return null;
 				} else {
-					return assetRepository.copyRawAsset(handle, fileHandle);
+					AssetRepository.getInstance().moveFile(source, target, false);
+					onComplete.run();
 				}
 			}
 		});
 	}
 
-	private static void handlePayloadMove (GlobalDragAndDrop.BaseDragAndDropPayload payloadObject, Item targetItem) {
-		if (payloadObject instanceof GlobalDragAndDrop.GameAssetDragAndDropPayload) {
-			//Do single drag and drop asset move
-			GameAsset<?> gameAsset = ((GlobalDragAndDrop.GameAssetDragAndDropPayload)payloadObject).getGameAsset();
-			FileHandle handle = gameAsset.getRootRawAsset().handle;
-			if (!handle.path().equals(targetItem.fileHandle.path())) {
-				AssetImporter.moveFile(handle, targetItem.fileHandle, false);
-			}
-		} else if (payloadObject instanceof GlobalDragAndDrop.FileHandleDragAndDropPayload) {
-			FileHandle handle = ((GlobalDragAndDrop.FileHandleDragAndDropPayload)payloadObject).getHandle();
-			if (!handle.path().equals(targetItem.fileHandle.path())) {
-				AssetImporter.moveFile(handle, targetItem.fileHandle, false);
-			}
-		} else if (payloadObject instanceof GlobalDragAndDrop.ArrayDragAndDropPayload) {
-			Array<GlobalDragAndDrop.BaseDragAndDropPayload> items = ((GlobalDragAndDrop.ArrayDragAndDropPayload)payloadObject).getItems();
-			Array<GlobalDragAndDrop.BaseDragAndDropPayload> itemCopy = new Array<>();
-			itemCopy.addAll(items);
-			for (int i = 0; i < itemCopy.size; i++) {
-				handlePayloadMove(itemCopy.get(i), targetItem);
+	private static boolean isDirectory (Item item) {
+		if (item == null || item.fileHandle == null) {
+			return false;
+		}
+		return item.fileHandle.isDirectory();
+	}
+
+	/**
+	 * Broken assets still belong to project.
+	 */
+	private static boolean belongsToProject (GlobalDragAndDrop.BaseDragAndDropPayload payload) {
+		if (payload instanceof GlobalDragAndDrop.FileHandleDragAndDropPayload) {
+			return belongsToProject((GlobalDragAndDrop.FileHandleDragAndDropPayload) payload);
+		} else if (payload instanceof GlobalDragAndDrop.GameAssetDragAndDropPayload) {
+			return belongsToProject((GlobalDragAndDrop.GameAssetDragAndDropPayload) payload);
+		} else if (payload instanceof GlobalDragAndDrop.ArrayDragAndDropPayload) {
+			return belongToProject((GlobalDragAndDrop.ArrayDragAndDropPayload) payload);
+		}
+		return false;
+	}
+
+	/**
+	 * Broken assets still belong to project.
+	 */
+	private static boolean belongsToProject (GlobalDragAndDrop.FileHandleDragAndDropPayload payload) {
+		if (payload == null || payload.getHandle() == null) {
+			return false;
+		}
+		String projectPath = SharedResources.currentProject.rootProjectDir().path();
+		String payloadPath = payload.getHandle().path();
+		return payloadPath.startsWith(projectPath);
+	}
+
+	/**
+	 * Broken assets still belong to project.
+	 */
+	private static boolean belongsToProject (GlobalDragAndDrop.GameAssetDragAndDropPayload payload) {
```

Comment:
Nice implementation, BUT i think we should change it. The only case where the asset is not belonging to the project is the native drag and drop, which is from one specific location and we know 100% that its not from the project. I think we should just have a proeprty on payload that lets us know if its "internal or external", its true by default, and in the case of native drag and drop we mark it.  

Doing this stuff on comparisons of paths isn't super nice, and if assets get broken, or there are dummy assets, this breaks because we are searchng for backing raw assets and looking at their handles

Reply:
Fixed